internal/event/event.go

```diff
@@ -148,8 +210,204 @@ func (ev Event) Application() (*v1alpha1.Application, error) {
+}
+
+func ResourceID(ev *cloudevents.Event) string {
+	id, ok := ev.Extensions()[resourceID].(string)
+	if ok {
+		return id
+	}
+
+	return ""
+}
+
+func EventID(ev *cloudevents.Event) string {
+	id, ok := ev.Extensions()[eventID].(string)
+	if ok {
+		return id
+	}
+
+	return ""
+}
+
 func (ev Event) AppProject() (*v1alpha1.AppProject, error) {
 	proj := &v1alpha1.AppProject{}
 	err := ev.event.DataAs(proj)
 	return proj, err
 }
+
+type streamWriter interface {
+	Send(*eventstreamapi.Event) error
+	Context() context.Context
+}
+
+// EventWriter keeps track of the latest event for resources and sends them on a given gRPC stream.
+// It resends the event with exponential backoff until the event is ACK'd and removed from its list.
+type EventWriter struct {
+	mu sync.RWMutex
+
+	// key: resource name + UID
+	// value: latest event for a resource
+	// - acquire 'lock' before accessing
+	latestEvents map[string]*eventMessage
+
+	// target refers to the specified gRPC stream.
+	target streamWriter
+
+	log *logrus.Entry
+}
+
+type eventMessage struct {
+	mu sync.RWMutex
+
+	// latest event for a resource
+	// - acquire 'lock' before accessing
+	event *cloudevents.Event
+
+	// retry sending the event after this time
+	retryAfter *time.Time
+
+	// config for exponential backoff
+	backoff *wait.Backoff
+}
+
+func NewEventWriter(target streamWriter) *EventWriter {
+	return &EventWriter{
+		latestEvents: map[string]*eventMessage{},
+		target:       target,
+		log: logrus.WithFields(logrus.Fields{
+			"module":      "EventWriter",
+			"client_addr": grpcutil.AddressFromContext(target.Context()),
+		}),
+	}
+}
+
+func (ew *EventWriter) Add(ev *cloudevents.Event) {
+	resID := ResourceID(ev)
+
+	ew.mu.Lock()
+	defer ew.mu.Unlock()
+
+	defaultBackoff := wait.Backoff{
+		Steps:    5,
+		Duration: 5 * time.Second,
+		Factor:   2.0,
+		Jitter:   0.1,
+	}
+
+	eventMsg, exists := ew.latestEvents[resID]
+	if !exists {
+		ew.latestEvents[resID] = &eventMessage{
+			event:   ev,
+			backoff: &defaultBackoff,
+		}
+		ew.log.Tracef("added a new event resourceID %s eventID %s", resID, EventID(ev))
+		return
+	}
+
+	// Replace the old event and reset the backoff.
+	eventMsg.mu.Lock()
+	eventMsg.event = ev
+	eventMsg.backoff = &defaultBackoff
+	eventMsg.retryAfter = nil
+	eventMsg.mu.Unlock()
+	ew.log.Tracef("updated an existing event: resourceID %s eventID %s", resID, EventID(ev))
+}
+
+func (ew *EventWriter) Get(resID string) *eventMessage {
+	ew.mu.RLock()
+	defer ew.mu.RUnlock()
+	return ew.latestEvents[resID]
+}
+
+func (ew *EventWriter) Remove(ev *cloudevents.Event) {
+	ew.mu.Lock()
+	defer ew.mu.Unlock()
+
+	// Remove the event only if it matches both the resourceID and eventID.
+	resourceID := ResourceID(ev)
+	latestEvent, exists := ew.latestEvents[resourceID]
+	if !exists {
+		return
+	}
+
+	if EventID(latestEvent.event) == EventID(ev) {
+		delete(ew.latestEvents, resourceID)
+	}
+}
+
+func (ew *EventWriter) SendWaitingEvents(ctx context.Context) {
```

Comment:
Should probably be noted that this function will never return unless the context is done, and therefore should be started in a go routine.